python/pyarrow/_fs.pyx

```diff
@@ -0,0 +1,522 @@
+    File = <int8_t> CFileType_File
+    Directory = <int8_t> CFileType_Directory
+
+
+cdef class FileStats:
+    """FileSystem entry stats"""
+
+    cdef CFileStats stats
+
+    def __init__(self):
+        raise TypeError('dont initialize me')
+
+    @staticmethod
+    cdef FileStats wrap(CFileStats stats):
+        cdef FileStats self = FileStats.__new__(FileStats)
+        self.stats = stats
+        return self
+
+    @property
+    def type(self):
+        """Type of the file
+
+        The returned enum variants have the following meanings:
+        - FileType.NonExistent: target does not exist
+        - FileType.Unknown: target exists but its type is unknown (could be a
+                            special file such as a Unix socket or character
+                            device, or Windows NUL / CON / ...)
+        - FileType.File: target is a regular file
+        - FileType.Directory: target is a regular directory
+
+        Returns
+        -------
+        type : FileType
+        """
+        return FileType(<int8_t> self.stats.type())
+
+    @property
+    def path(self):
+        """The full file path in the filesystem."""
+        return pathlib.Path(frombytes(self.stats.path()))
+
+    @property
+    def base_name(self):
+        """The file base name
+
+        Component after the last directory separator.
+        """
+        return frombytes(self.stats.base_name())
+
+    @property
+    def size(self):
+        """The size in bytes, if available
+
+        Only regular files are guaranteed to have a size.
+        """
+        if self.stats.type() != CFileType_File:
+            raise ValueError(
+                'Only regular files are guaranteed to have a size'
+            )
+        return self.stats.size()
+
+    @property
+    def extension(self):
+        """The file extension"""
+        return frombytes(self.stats.extension())
+
+    @property
+    def mtime(self):
+        """The time of last modification, if available.
+
+        Returns
+        -------
+        mtime : datetime.datetime
+        """
+        cdef PyObject *out
+        check_status(PyDateTime_from_TimePoint(self.stats.mtime(), &out))
+        return PyObject_to_object(out)
+
+
+cdef class Selector:
+    """File and directory selector.
+
+    It containt a set of options that describes how to search for files and
+    directories.
+
+    Parameters
+    ----------
+    base_dir : str or pathlib.Path
+        The directory in which to select files.
+        If the path exists but doesn't point to a directory, this should be an
+        error.
+    allow_non_existent : bool, default False
+        The behavior if `base_dir` doesn't exist in the filesystem.
+        If false, an error is returned.
+        If true, an empty selection is returned.
+    recursive : bool, default False
+        Whether to recurse into subdirectories.
+    """
+    cdef CSelector selector
+
+    def __init__(self, base_dir='', bint allow_non_existent=False,
```

Comment:
For LocalFileSystem `''` means the root, `'.'` fails., 
For SubTreeFileSystem `''` and `'.'` are equivalent.

I suppose this is because path concatenation. cc @pitrou

Reply:
`'.'` is not supported currently indeed.

Reply:
By the way, having a default value for `base_dir` doesn't make much sense. You'll probably always specify an explicit base directory.

Reply:
`'.'` seems to work for me:

```
In [3]: s = pa.fs.Selector('.') 

In [4]: fs = pa.fs.LocalFileSystem() 

In [5]: fs.get_target_stats(s)  
Out[5]: 
[<pyarrow._fs.FileStats at 0x7f531c7a1c10>,
 <pyarrow._fs.FileStats at 0x7f531c7a1850>,
 <pyarrow._fs.FileStats at 0x7f531c7a1120>,
...

In [6]: fs.get_target_stats(s)[0].path 
Out[6]: PosixPath('Untitled2.ipynb')
```

where `Untitled2.ipynb` is a file of the current working directory of my python process

Reply:
Will remove the default tomorrow.